lib/Ast.ml

```diff
@@ -1,223 +0,0 @@
-
-and prefixop = Not | UMinus | UPlus
-
-and postfixop = Transpose
-
-and 'e index =
-  | All
-  | Single of 'e
-  | Upfrom of 'e
-  | Downfrom of 'e
-  | Between of 'e * 'e
-  | Multiple of 'e
-
-and 'e expression =
-  | Conditional of 'e * 'e * 'e
-  | InfixOp of 'e * infixop * 'e
-  | PrefixOp of prefixop * 'e
-  | PostfixOp of 'e * postfixop
-  | Variable of identifier
-  | IntNumeral of string
-  | RealNumeral of string
-  | FunApp of identifier * 'e list
-  | CondFunApp of identifier * 'e list
-  | GetLP
-  (* deprecated *)
-  | GetTarget
-  | ArrayExpr of 'e list
-  | RowVectorExpr of 'e list
-  | Paren of 'e
-  | Indexed of 'e * 'e index list
-
-and expression_untyped_metadata =
-  {expr_untyped_meta_loc: location sexp_opaque [@compare.ignore]}
-
-and expression_typed_metadata =
-  { expr_typed_meta_origin_type: originblock * unsizedtype
-  ; expr_typed_meta_loc: location sexp_opaque [@compare.ignore] }
-
-and untyped_expression =
-  | UntypedExpr of (untyped_expression expression * expression_untyped_metadata)
-
-and typed_expression =
-  | TypedExpr of (typed_expression expression * expression_typed_metadata)
-
-(* == Statements == *)
-and assignmentoperator =
-  | Assign
-  | PlusAssign
-  | MinusAssign
-  | TimesAssign
-  | DivideAssign
-  | EltTimesAssign
-  | EltDivideAssign
-  | ArrowAssign
-
-(* deprecated *)
-and 'e truncation =
-  | NoTruncate
-  | TruncateUpFrom of 'e
-  | TruncateDownFrom of 'e
-  | TruncateBetween of 'e * 'e
-
-and 'e printable = PString of string | PExpr of 'e
-
-and 'e sizedtype =
-  | SInt
-  | SReal
-  | SVector of 'e
-  | SRowVector of 'e
-  | SMatrix of 'e * 'e
-  | SArray of 'e sizedtype * 'e
-
-and 'e transformation =
-  | Identity
-  | Lower of 'e
-  | Upper of 'e
-  | LowerUpper of 'e * 'e
-  | LocationScale of 'e * 'e
-  | Ordered
-  | PositiveOrdered
-  | Simplex
-  | UnitVector
-  | CholeskyCorr
-  | CholeskyCov
-  | Correlation
-  | Covariance
-
-and ('e, 's) statement =
-  | Assignment of
-      { assign_identifier: identifier
-      ; assign_indices: 'e index list
-      ; assign_op: assignmentoperator
-      ; assign_rhs: 'e }
-  | NRFunApp of identifier * 'e list
-  | TargetPE of 'e
-  | IncrementLogProb of 'e
-  (* deprecated *)
-  | Tilde of
-      { arg: 'e
-      ; distribution: identifier
-      ; args: 'e list
-      ; truncation: 'e truncation }
-  | Break
-  | Continue
-  | Return of 'e
-  | ReturnVoid
-  | Print of 'e printable list
-  | Reject of 'e printable list
-  | Skip
-  | IfThenElse of 'e * 's * 's
-  | IfThen of 'e * 's
-  | While of 'e * 's
-  | For of
-      { loop_variable: identifier
-      ; lower_bound: 'e
-      ; upper_bound: 'e
-      ; loop_body: 's }
-  | ForEach of identifier * 'e * 's
-  (* TODO: maybe make separate nodes for matrix and array foreach loops? *)
-  | Block of 's list
-  | VDecl of 'e sizedtype * identifier
-  | VDeclAss of {sizedtype: 'e sizedtype; identifier: identifier; value: 'e}
-  | TVDecl of 'e sizedtype * 'e transformation * identifier
```

Comment:
T for Top could probably be changed to be a little more descriptive - maybe constrainable or even global? not sure

Reply:
I'm also still wondering if we would be served just as well with a single Decl that encapsulated all the possibilities orthogonally:
```
VDecl of {
sizedtype: 'e sizedtype;
id: identifier;
value: 'e option;
transformation: 'e transformation option;
write_out: boolean;
global: boolean;
}
```
Or something like that? My reasoning is around the idea that we might eventually want to orthogonally toggle whether we write a variable out or not, and that the code for generating these things would probably be easier in this form (though we'll be working from the MIR). But maybe the semantic analysis is better in the current form?

Reply:
You may be right. I could be convinced. It would require some refactoring though.

Not sure how the write_out flag would work in code_gen. Could you figure that out? Should I already add it?

It seems to me like we only ever want to write out global though, right? So then, in a sense, we'd like the write_out flag only to occur for globals?

Also, do we need the globals flag? Isn't something with a transformation simply the same as a global var?

Reply:
This has been fixed now.